pyomo/contrib/mindtpy/mip_solve.py

```diff
@@ -2,17 +2,280 @@
 from __future__ import division
 
 from pyomo.contrib.gdpopt.util import copy_var_list_values
-from pyomo.core import Constraint, Expression, Objective, minimize, value
+from pyomo.core import Constraint, Expression, Objective, minimize, value, Var
 from pyomo.opt import TerminationCondition as tc
 from pyomo.opt import SolutionStatus, SolverFactory
 from pyomo.contrib.gdpopt.util import SuppressInfeasibleWarning, _DoNothing
 from pyomo.contrib.gdpopt.mip_solve import distinguish_mip_infeasible_or_unbounded
+from pyomo.solvers.plugins.solvers.persistent_solver import PersistentSolver
+
+from pyomo.contrib.mindtpy.nlp_solve import (solve_NLP_subproblem,
+                                             handle_NLP_subproblem_optimal, handle_NLP_subproblem_infeasible,
+                                             handle_NLP_subproblem_other_termination, solve_NLP_feas)
+from pyomo.contrib.mindtpy.cut_generation import (add_oa_cuts,
+                                                  add_int_cut)
+from pyomo.contrib.gdpopt.util import copy_var_list_values, identify_variables
+from math import copysign
+from pyomo.environ import *
+from pyomo.core import Constraint, minimize, value
+from pyomo.core.expr import current as EXPR
+from math import fabs
+
+from pyomo.repn import generate_standard_repn
+
+try:
+    import cplex
+    from cplex.callbacks import LazyConstraintCallback
+except ImportError:
+    print("Cplex python API is not found. Therefore, lp-nlp is not supported")
+    '''Other solvers (e.g. Gurobi) are not supported yet'''
```

Comment:
Please do not print your message here.  The way this is currently implemented, this message is printed out *every* time `pyomo.environ` is imported on platforms that do not have cplex.  Also, messages should be emitted using the logger and not with `print()`.  

Look at `pyomo.common.dependencies.attempt_import()` for an alternative approach for working with optional dependencies.  You should be able to just:
```python
cplex, cplex_available = attempt_import('cplex')
```

More generally, you need to re-think either how you define `LazyOACallback`, or else when/where this module is imported.  Because you automatically import this module, your current implementation causes a fatal exception that prevents importing `pyomo.environ` if the cplex python library is not present (see the test failures).

Reply:
I created a new file name **single_tree.py** and moved the `LazyOACallback` class to it.
I also used `attempt_import` to import both `cplex` and `pyomo.contrib.mindtpy.single_tree`.
Please let me know if `attempt_import` is inappropriately used or related code can be further improved.

Reply:
Hi John. I pushed several commits after the pull request. Now there is an **Install dependencies** bug on several tests. Do you know how to fix it? Thank you. @jsiirola

Reply:
Hi John. Do you know why the test `continuous-integration/jenkins/pr` failed? It seems a Jenkins failure.
Thank you.
@jsiirola